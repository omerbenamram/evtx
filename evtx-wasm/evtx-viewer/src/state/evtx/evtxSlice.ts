import type { EvtxFileInfo } from "../../lib/types";

// ---------------- Types & State ----------------
export interface EvtxMetaState {
  isLoading: boolean;
  loadingMessage: string;
  matchedCount: number;
  totalRecords: number;
  fileInfo: EvtxFileInfo | null;
  currentFileId: string | null;
}

export const evtxInitialState: EvtxMetaState = {
  isLoading: false,
  loadingMessage: "",
  matchedCount: 0,
  totalRecords: 0,
  fileInfo: null,
  currentFileId: null,
};

// ---------------- Action Types ----------------
export type EvtxAction = {
  type: "evtx/UPDATE";
  payload: Partial<EvtxMetaState>;
};

// ---------------- Reducer ----------------
export function evtxReducer(
  state: EvtxMetaState = evtxInitialState,
  action: EvtxAction
): EvtxMetaState {
  switch (action.type) {
    case "evtx/UPDATE":
      return { ...state, ...action.payload };
    default:
      return state;
  }
}

// ---------------- Action Creators ----------------
export const updateEvtxMeta = (
  payload: Partial<EvtxMetaState>
): EvtxAction => ({
  type: "evtx/UPDATE",
  payload,
});
