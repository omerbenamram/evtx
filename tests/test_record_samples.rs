mod fixtures;
use fixtures::*;

use evtx::{EvtxParser, ParserSettings};
use serde_json::Value;

#[test]
fn test_event_xml_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_json_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_xml_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_json_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_xml_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_xml_sample_with_user_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_xml_sample_with_entity_ref() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(record.data);
}

#[test]
fn test_event_xml_sample_with_entity_ref_2() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 25)
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(record.data);
}

#[test]
fn test_event_json_with_multiple_nodes_same_name() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_sample_with_separate_json_attributes() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/Application.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_with_multiple_data_elements() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/MSExchange_Management_wec.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record_xml = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(&first_record_xml.data);

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_with_multiple_nodes_same_name_separate() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_multiple_empty_data_nodes_not_ignored() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/issue_201.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .next()
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}
