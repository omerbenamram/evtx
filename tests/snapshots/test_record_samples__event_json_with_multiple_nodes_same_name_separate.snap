---
source: tests/test_record_samples.rs
assertion_line: 258
expression: "&value"
---
{
  "Event": {
    "System": {
      "Channel": "Microsoft-Windows-CAPI2/Operational",
      "Computer": "WIN-M5327EF98B9",
      "Correlation": null,
      "EventID": 53,
      "EventRecordID": 28,
      "Execution_attributes": {
        "ProcessID": 1396,
        "ThreadID": 2132
      },
      "Keywords": "0x4000000000000036",
      "Level": 4,
      "Opcode": 2,
      "Provider_attributes": {
        "Guid": "{5bbca4a8-b209-48dc-a8c7-b23d3e5216fb}",
        "Name": "Microsoft-Windows-CAPI2"
      },
      "Security_attributes": {
        "UserID": "S-1-5-21-1223297778-3299746493-1462173606-500"
      },
      "Task": 53,
      "TimeCreated_attributes": {
        "SystemTime": "2017-05-19T02:02:36.203125Z"
      },
      "Version": 0
    },
    "UserData": {
      "CryptRetrieveObjectByUrlWire": {
        "AdditionalInfo": {
          "Action_attributes": {
            "name": "NoProxy"
          },
          "HTTPRequestHeadersInfo": {
            "Header": "Connection: Keep-Alive",
            "Header_1": "GET /pki/crl/products/microsoftrootcert.crl HTTP/1.1",
            "Header_2": "Accept: */*",
            "Header_3": "If-None-Match: \"ea9ee7b1bc43d21:0\"",
            "Header_4": "If-Modified-Since: Mon, 21 Nov 2016 06:01:26 GMT",
            "Header_5": "Cache-Control: max-age = 900",
            "Header_6": "User-Agent: Microsoft-CryptoAPI/6.1"
          },
          "HTTPResponseHeadersInfo": {
            "Header": "x-ms-blob-type: BlockBlob",
            "Header_1": "HTTP/1.1 200 OK",
            "Header_10": "x-ms-version: 2009-09-19",
            "Header_11": "x-ms-lease-status: unlocked",
            "Header_2": "Connection: keep-alive",
            "Header_3": "Date: Thu, 18 May 2017 11:37:58 GMT",
            "Header_4": "Content-Length: 813",
            "Header_5": "Content-Type: application/pkix-crl",
            "Header_6": "Last-Modified: Tue, 02 May 2017 22:24:24 GMT",
            "Header_7": "ETag: 0x8D491A9FD112A27",
            "Header_8": "Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0",
            "Header_9": "x-ms-request-id: 477c132d-0001-0045-443b-c49ae1000000"
          },
          "NetworkConnectivityStatus_attributes": {
            "_SENSAPI_NETWORK_ALIVE_LAN": "true",
            "value": "1"
          }
        },
        "AuxInfo_attributes": {
          "fProxyCacheRetrieval": "true",
          "maxUrlRetrievalByteCount": "104857600"
        },
        "CacheInfo": {
          "URLCacheResponseInfo_attributes": {
            "lastModifiedTime": "2017-05-02T22:24:24Z",
            "responseType": "CRYPTNET_URL_CACHE_RESPONSE_HTTP"
          }
        },
        "CacheInfo_attributes": {
          "lastSyncTime": "2017-05-19T02:02:36.203Z"
        },
        "CorrelationAuxInfo_attributes": {
          "SeqNumber": "7",
          "TaskId": "{74E4CD40-C966-49F5-B50A-032DFFEE57CA}"
        },
        "EventAuxInfo_attributes": {
          "ProcessName": "Setup.exe"
        },
        "Flags_attributes": {
          "CRYPT_LDAP_SCOPE_BASE_ONLY_RETRIEVAL": "true",
          "CRYPT_PROXY_CACHE_RETRIEVAL": "true",
          "CRYPT_RETRIEVE_MULTIPLE_OBJECTS": "true",
          "CRYPT_WIRE_ONLY_RETRIEVAL": "true",
          "value": "202005"
        },
        "Object_attributes": {
          "constant": "2",
          "type": "CONTEXT_OID_CRL"
        },
        "Result_attributes": {
          "value": "0"
        },
        "RetrievedObjects": {
          "CertificateRevocationList_attributes": {
            "fileRef": "0986764ED95D3C77F3F1AD8340EDD2F36C3BF8E7.crl",
            "issuerName": "Microsoft Root Certificate Authority"
          }
        },
        "Timeout": "PT9.984S",
        "URL": "http://crl.microsoft.com/pki/crl/products/microsoftrootcert.crl",
        "URL_attributes": {
          "scheme": "http"
        }
      }
    }
  },
  "Event_attributes": {
    "xmlns": "http://schemas.microsoft.com/win/2004/08/events/event"
  }
}
